indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -710,20 +728,8 @@ private static DimensionsSpec createDimensionsSpec(List<Pair<QueryableIndex, Dat
     // Dimensions are extracted from the recent segments to olders because recent segments are likely to be queried more
     // frequently, and thus the performance should be optimized for recent ones rather than old ones.
 
-    // timelineSegments are sorted in order of interval, but we do a sanity check here.
-    final Comparator<Interval> intervalComparator = Comparators.intervalsByStartThenEnd();
-    for (int i = 0; i < queryableIndices.size() - 1; i++) {
-      final Interval shouldBeSmaller = queryableIndices.get(i).lhs.getDataInterval();
-      final Interval shouldBeLarger = queryableIndices.get(i + 1).lhs.getDataInterval();
-      Preconditions.checkState(
-          intervalComparator.compare(shouldBeSmaller, shouldBeLarger) <= 0,
-          "QueryableIndexes are not sorted! Interval[%s] of segment[%s] is laster than interval[%s] of segment[%s]",
-          shouldBeSmaller,
-          queryableIndices.get(i).rhs.getId(),
-          shouldBeLarger,
-          queryableIndices.get(i + 1).rhs.getId()
-      );
-    }
+    // sort timelineSegments in order of interval.
+    queryableIndices.sort((o1, o2) -> Comparators.intervalsByStartThenEnd().compare(o1.rhs.getInterval(), o2.rhs.getInterval()));
```

Comment:
Would you explain why you changed this? The intention here was to make some unit tests failed when the timeline somehow returns unsorted intervals which is supposed to be a bug.

Reply:
Hi @jihoonson 

Here I intended to fix the error in step 5 as described [here](https://github.com/apache/druid/issues/9904).

When compacting the two segments

- `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`
- `inline_data_2020-05-18T20:00:00.000Z_2020-05-18T21:00:00.000Z_version_2`

there will be three timelineObjectHolders

- `2020-05-18T00:00:00.000Z/2020-05-18T20:00:00.000Z` backed by segment `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`
- `2020-05-18T20:00:00.000Z/2020-05-18T21:00:00.000Z` backed by segment `inline_data_2020-05-18T20:00:00.000Z_2020-05-18T21:00:00.000Z_version_2`
- `2020-05-18T21:00:00.000Z/2020-05-19T00:00:00.000Z` backed by segment `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`

And the last two pairs of `queryableIndices` will fail to pass the previous sanity check. Also I found that now timeline already has unit tests about sorting timelineObjectHolders and the sanity check will pass in most case. So I decided to fix the error by re-sorting them anyway. And I added a link in the comment. Does it make sense?

Reply:
Hi @jihoonson 

Here I intended to fix the error in step 5 as described [here](https://github.com/apache/druid/issues/9904). 

When compacting the two segments

- `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`
- `inline_data_2020-05-18T20:00:00.000Z_2020-05-18T21:00:00.000Z_version_2`

there will be three timelineObjectHolders

- `2020-05-18T00:00:00.000Z/2020-05-18T20:00:00.000Z` backed by segment `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`
- `2020-05-18T20:00:00.000Z/2020-05-18T21:00:00.000Z` backed by segment `inline_data_2020-05-18T20:00:00.000Z_2020-05-18T21:00:00.000Z_version_2`
- `2020-05-18T21:00:00.000Z/2020-05-19T00:00:00.000Z` backed by segment `inline_data_2020-05-18T00:00:00.000Z_2020-05-19T00:00:00.000Z_version_1`

And the last two pairs of `queryableIndices` will fail to pass the previous sanity check. Also I found that now timeline already has unit tests about sorting timelineObjectHolders and the sanity check will pass in most case. So I decided to fix the error by re-sorting them anyway. And I added a link in the comment. Do it make sense?

Reply:
@yuanlihan thanks for the explanation! What this method does is finding the most reasonable order of dimensions when input segments have mixed orders. The current algorithm is assuming that more recent segments will likely have the order what you want more (there would be some reason if you have changed the dimension order at some point). Maybe a potential improvement can be considering the segment version as well if some segment intervals are overlapped, but I think this could be done in a separate PR.

> Also I found that now timeline already has unit tests about sorting timelineObjectHolders and the sanity check will pass in most case.

BTW, out of curiosity, can you point me out where this test is?

Reply:
Hi @jihoonson 

> BTW, out of curiosity, can you point me out where this test is?

I found that the expected timelines in test cases of `VersionedIntervalTimelineTest` are all intentionally provided in well-ordered. Maybe I should said that there are implicit tests that cover sorting timelineObjectHolders.